source/server/admin/stats_handler.cc

```diff
@@ -265,16 +331,90 @@ StatsHandler::statsAsJson(const std::map<std::string, uint64_t>& all_stats,
+          histogram->intervalStatistics().supportedBuckets(),
+          histogram->intervalStatistics().computedBuckets(),
+          histogram->cumulativeStatistics().computedBuckets(), histogram->name()));
+    }
+  }
+  if (!histogram_obj_array.empty()) { // If used histogram found.
+    histograms_obj_container_fields["histograms"] = ValueUtil::listValue(histogram_obj_array);
+  }
+}
 
-  return MessageUtil::getJsonStringFromMessageOrDie(document, pretty_print, true);
+void StatsHandler::statsAsJsonDisjointHistogramBucketsHelper(
+    Protobuf::Map<std::string, ProtobufWkt::Value>& histograms_obj_container_fields,
+    const std::vector<Stats::ParentHistogramSharedPtr>& all_histograms, bool used_only,
+    const absl::optional<std::regex>& regex) {
+  std::vector<ProtobufWkt::Value> histogram_obj_array;
+  for (const Stats::ParentHistogramSharedPtr& histogram : all_histograms) {
+    if (shouldShowMetric(*histogram, used_only, regex)) {
+      histogram_obj_array.push_back(statsAsJsonHistogramBucketsCreateHistogramElementHelper(
+          histogram->intervalStatistics().supportedBuckets(),
+          histogram->intervalStatistics().computeDisjointBuckets(),
+          histogram->cumulativeStatistics().computeDisjointBuckets(), histogram->name()));
+    }
+  }
+  if (!histogram_obj_array.empty()) { // If used histogram found.
+    histograms_obj_container_fields["histograms"] = ValueUtil::listValue(histogram_obj_array);
+  }
+}
+
+ProtobufWkt::Value StatsHandler::statsAsJsonHistogramBucketsCreateHistogramElementHelper(
+    Stats::ConstSupportedBuckets& supported_buckets, const std::vector<uint64_t>& interval_buckets,
+    const std::vector<uint64_t>& cumulative_buckets, const std::string& name) {
+  ProtobufWkt::Struct histogram_obj;
+  auto* histogram_obj_fields = histogram_obj.mutable_fields();
+  (*histogram_obj_fields)["name"] = ValueUtil::stringValue(name);
+
+  // Make sure all vectors are the same size.
+  ASSERT(interval_buckets.size() == cumulative_buckets.size());
+  ASSERT(cumulative_buckets.size() == supported_buckets.size());
+  size_t min_size =
+      std::min({interval_buckets.size(), cumulative_buckets.size(), supported_buckets.size()});
+
+  std::vector<ProtobufWkt::Value> supported_bucket_array;
+  std::vector<ProtobufWkt::Value> computed_bucket_array;
+  for (size_t i = 0; i < min_size; ++i) {
+    supported_bucket_array.push_back(ValueUtil::numberValue(supported_buckets[i]));
+
+    ProtobufWkt::Struct computed_bucket;
+    auto* computed_bucket_fields = computed_bucket.mutable_fields();
+    (*computed_bucket_fields)["interval"] = ValueUtil::numberValue(interval_buckets[i]);
+    (*computed_bucket_fields)["cumulative"] = ValueUtil::numberValue(cumulative_buckets[i]);
+    computed_bucket_array.push_back(ValueUtil::structValue(computed_bucket));
+  }
+  (*histogram_obj_fields)["supported_buckets"] = ValueUtil::listValue(supported_bucket_array);
```

Comment:
I think this should be named "configured_buckets" or "bucket_upper_bounds" or something like that. But on the other hand, we already have "supported_quantiles" in the output. But I don't find that "supported" is very clear about what it is. @jmarantz wdyt?

Reply:
I combined the arrays and the combined array has been renamed to "buckets".